BUG: stats.tukeylambda: Fix the support() method.

WarrenWeckesser opened the pull request:
Closes gh-21369.

WarrenWeckesser committed:
BUG: stats.tukeylambda: Fix the support() method.

Closes gh-21369.

WarrenWeckesser commented:
There is a new unrelated test failure on github actions: `Windows tests / fail slow, full, py3.10/npMin, dev.py`. It is the sparse linalg test `TestExpmActionInterval.test_expm_multiply_interval_matrix`

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -10481,12 +10481,20 @@ class tukeylambda_gen(rv_continuous):
     %(example)s
 
     """
+    _support_mask = rv_continuous._open_support_mask
```
~~I looked up what this does. If I'm thinking straight, this should only have an effect on the output of the `pdf` method, right? And if I can extrapolate from the picture on Wikipedia, it looks like the PDF at `x=a` and `x=b` will always be zero, so unless the PDF was not returning the right answer at the endpoints, this would not have an effect. Did you change this just to reduce the calculation for these special cases, or does it fix something?~~ I see you did this to avoid a warning.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -10481,12 +10481,20 @@ class tukeylambda_gen(rv_continuous):
         return [_ShapeInfo("lam", False, (-np.inf, np.inf), (False, False))]
 
+    def _get_support(self, lam):
+        b = _lazywhere(lam > 0, (lam,),
+                       f=lambda lam: 1/lam,
+                       f2=lambda lam: np.inf)
```
Just to show that I'm actually thinking about this - not that there's any real motivation to change this - it would be slightly _simpler_ as:
```suggestion
        b = _lazywhere(lam <= 0, (lam,), np.inf, f2=1/lam)
```
right?
And `_argcheck` should already be taking care of when `lam` is invalid, but I suppose only this way would do the right thing if `lam = np.nan`.

WarrenWeckesser replied:
That's not quite right, but you are right that it can be simplified: replace `f2=lambda lam: np.inf` with `fillvalue=np.inf`.  I'll push that change.

mdhaber replied:
Oops, should have checked the signature. OK, I'll merge after that.

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -7870,6 +7870,22 @@ def test_regression_tukey_lambda():
     assert_((p[2] == 0.0).any())
 
 
+def test_tukeylambda_support():
```
I'd be willing to review a commit that brought all the scattered `tukeylambda` tests into one class.

WarrenWeckesser replied:
I almost did that originally, so I'll do that now.  I'm seeing some quirks (possible bugs) in `test_regression_tukey_lambda` that I'll clean up while I'm at it.

WarrenWeckesser replied:
The bug is the test `assert_(~np.isnan(p).all())`.  The intent was almost certainly to check that there are no nans in `p`, but all it takes for that test to succeed is at least one non-nan value in `p`.  I'll change it to `assert np.isfinite(p).all()`.

mdhaber reviewed (approved):
Looks fine; I guess I'd just ask whether you'd like to add a commit that puts all the `tukeylambda` tests in one class. (They are scattered throughout the file right now.) It might also be nice to document the support as most distributions do, but I know the documentation of `tukeylambda` is already a bit weird since it doesn't have a simple PDF.

WarrenWeckesser committed:
MAINT: Simplify a use of _lazywhere.

WarrenWeckesser committed:
MAINT: stats: Clean up tukeylambda tests.

Test cleanup:
* Move existing tests to a class.
* Split an old test into two tests, one of which
  is parametrized.
* Fix a bug in the check for lack of nans in the
  result of the pdf method.
* Add comments to clarify what the tests are doing

Also fix a spurious warning that was generated by
tukeylambda.pdf when lam was 0.

WarrenWeckesser committed:
MAINT: stats: Move one more test to the TestTukeyLambda class.
